File: LeetCode/37.sudoku-solver.py
#
# @lc app=leetcode id=37 lang=python3
#
# [37] Sudoku Solver
#
# https://leetcode.com/problems/sudoku-solver/description/
#
# algorithms
# Hard (63.71%)
# Likes:    10104
# Dislikes: 291
# Total Accepted:    751.5K
# Total Submissions: 1.2M
# Testcase Example:  '[["5","3",".",".","7",".",".",".","."],["6",".",".","1","9","5",".",".","."],[".","9","8",".",".",".",".","6","."],["8",".",".",".","6",".",".",".","3"],["4",".",".","8",".","3",".",".","1"],["7",".",".",".","2",".",".",".","6"],[".","6",".",".",".",".","2","8","."],[".",".",".","4","1","9",".",".","5"],[".",".",".",".","8",".",".","7","9"]]'
#
# Write a program to solve a Sudoku puzzle by filling the empty cells.
# 
# A sudoku solution must satisfy all of the following rules:
# 
# 
# Each of the digits 1-9 must occur exactly once in each row.
# Each of the digits 1-9 must occur exactly once in each column.
# Each of the digits 1-9 must occur exactly once in each of the 9 3x3 sub-boxes
# of the grid.
# 
# 
# The '.' character indicates empty cells.
# 
# 
# Example 1:
# 
# 
# Input: board =
# [["5","3",".",".","7",".",".",".","."],["6",".",".","1","9","5",".",".","."],[".","9","8",".",".",".",".","6","."],["8",".",".",".","6",".",".",".","3"],["4",".",".","8",".","3",".",".","1"],["7",".",".",".","2",".",".",".","6"],[".","6",".",".",".",".","2","8","."],[".",".",".","4","1","9",".",".","5"],[".",".",".",".","8",".",".","7","9"]]
# Output:
# [["5","3","4","6","7","8","9","1","2"],["6","7","2","1","9","5","3","4","8"],["1","9","8","3","4","2","5","6","7"],["8","5","9","7","6","1","4","2","3"],["4","2","6","8","5","3","7","9","1"],["7","1","3","9","2","4","8","5","6"],["9","6","1","5","3","7","2","8","4"],["2","8","7","4","1","9","6","3","5"],["3","4","5","2","8","6","1","7","9"]]
# Explanation: The input board is shown above and the only valid solution is
# shown below:
# 
# 
# 
# 
# 
# Constraints:
# 
# 
# board.length == 9
# board[i].length == 9
# board[i][j] is a digit or '.'.
# It is guaranteed that the input board has only one solution.
# 
# 
#

# @lc code=start
class Solution:
    def solveSudoku(self, board: List[List[str]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        def is_valid(row: int, col: int, num: str) -> bool:
            # 检查行是否有效
            for i in range(9):
                if board[row][i] == num:
                    return False
            # 检查列是否有效
            for i in range(9):
                if board[i][col] == num:
                    return False
            # 检查 3x3 子盒子是否有效
            box_row, box_col = 3 * (row // 3), 3 * (col // 3)
            for i in range(box_row, box_row + 3):
                for j in range(box_col, box_col + 3):
                    if board[i][j] == num:
                        return False
            return True

        def backtrack() -> bool:
            for i in range(9):
                for j in range(9):
                    if board[i][j] == '.':
                        for num in map(str, range(1, 10)):
                            if is_valid(i, j, num):
                                board[i][j] = num
                                if backtrack():
                                    return True
                                board[i][j] = '.'  # 回溯
                        return False
            return True

        backtrack()
# @lc code=end

File: 00LeetCodeCpp/37.sudoku-solver.cpp
/*
 * @lc app=leetcode id=37 lang=cpp
 *
 * [37] Sudoku Solver
 */

// @lc code=start
class Solution {
    // 三个记事本，记录数字是否被占用
    bool row[9][10] = {false};
    bool col[9][10] = {false};
    bool box[9][10] = {false};
public:
    void solveSudoku(vector<vector<char>>& board) {
        // 1. 先扫描一遍，把初始数字登记在案
        for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++) {
                if (board[i][j] != '.') {
                    int num = board[i][j] - '0';
                    row[i][num] = col[j][num] = box[(i/3)*3 + j/3][num] = true;
                }
            }
        }
        backtrack(board, 0, 0);
        // backtrack(board);
    }

    bool backtrack(vector<vector<char>>& board, int r, int c) {
        // 自动换行逻辑：走到列末尾就跳到下一行开头
        if (c == 9) return backtrack(board, r + 1, 0);
        if (r == 9) return true; // 全部填完
        if (board[r][c] != '.') return backtrack(board, r, c + 1);

        int b_idx = (r / 3) * 3 + (c / 3);
        for (int num = 1; num <= 9; num++) {
            // O(1) 检查：直接看记事本，不用去扫棋盘
            if (!row[r][num] && !col[c][num] && !box[b_idx][num]) {
                // 做选择
                board[r][c] = num + '0';
                row[r][num] = col[c][num] = box[b_idx][num] = true;

                if (backtrack(board, r, c + 1)) return true;

                // 回溯：擦掉选择，本子上也要改回来
                board[r][c] = '.';
                row[r][num] = col[c][num] = box[b_idx][num] = false;
            }
        }
        return false;
    }
// private:
//     bool backtrack(vector<vector<char>>& board) {
//         for (int i = 0; i < 9; i++) {
//             for (int j = 0; j < 9; j++) {
//                 if (board[i][j] != '.') continue; // 跳过已填好的

//                 for (char c = '1'; c <= '9'; c++) {
//                     if (isValid(board, i, j, c)) {
//                         board[i][j] = c; // 1. 做选择
//                         if (backtrack(board)) return true; // 2. 递归：如果成功了就一直返回 true
//                         board[i][j] = '.'; // 3. 回溯：失败了就撤销选择
//                     }
//                 }
//                 return false; // 1-9 都试过了不行，说明这条路死胡同
//             }
//         }
//         return true; // 所有格子都填完了
//     }

//     // 检查在 (row, col) 填入字符 c 是否合法
//     bool isValid(vector<vector<char>>& board, int r, int c, char val) {
//         for (int i = 0; i < 9; i++) {
//             if (board[r][i] == val) return false; // 查行
//             if (board[i][c] == val) return false; // 查列
//             if (board[(r/3)*3 + i/3][(c/3)*3 + i%3] == val) return false; // 查宫格
//         }
//         return true;
//     }
};
// @lc code=end

